core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombineSequence.java

```diff
@@ -0,0 +1,315 @@
+    final int numChildOfParentNodes = numCurLevelNodes + (numChildrenForLastNode == 1 ? 1 : 0);
+
+    if (numChildOfParentNodes == 1) {
+      return numCurLevelNodes;
+    } else {
+      return numCurLevelNodes +
+             computeNumRequiredThreads(numChildOfParentNodes, combineDegree);
+    }
+  }
+
+  private Pair<Sequence<T>, Future> runCombine(
+      ReferenceCountingResourceHolder processingThreadHolder,
+      List<Sequence<T>> sequenceList
+  )
+  {
+    final Sequence<? extends Sequence<T>> sequences = Sequences.simple(sequenceList);
+    final CombiningSequence<T> combiningSequence = CombiningSequence.create(
+        new MergeSequence<>(ordering, sequences),
+        ordering,
+        combineFn
+    );
+
+    final BlockingQueue<ValueHolder> queue = new ArrayBlockingQueue<>(queueSize);
+
+    final Future future = exec.submit(
+        new PrioritizedRunnable()
+        {
+          @Override
+          public int getPriority()
+          {
+            return queryPriority;
+          }
+
+          @Override
+          public void run()
+          {
+            try {
+              combiningSequence.accumulate(
+                  queue,
+                  (theQueue, v) -> {
+                    try {
+                      addToQueue(theQueue, new ValueHolder(v));
+                    }
+                    catch (InterruptedException e) {
+                      throw new RuntimeException(e);
+                    }
+                    return theQueue;
+                  }
+              );
+              // add a null to indicate this is the last one
+              addToQueue(queue, new ValueHolder(null));
+            }
+            catch (InterruptedException e) {
+              throw new RuntimeException(e);
+            }
+            finally {
+              processingThreadHolder.close();
+            }
+          }
+        }
+    );
+
+    final Sequence<T> backgroundCombineSequence = new BaseSequence<>(
+        new IteratorMaker<T, Iterator<T>>()
+        {
+          @Override
+          public Iterator<T> make()
+          {
+            return new Iterator<T>()
+            {
+              private T nextVal;
+
+              @Override
+              public boolean hasNext()
+              {
+                try {
+                  final ValueHolder holder;
+                  if (!hasTimeout) {
+                    holder = queue.take();
+                  } else {
+                    final long timeout = timeoutAt - System.currentTimeMillis();
+                    holder = queue.poll(timeout, TimeUnit.MILLISECONDS);
+                  }
+
+                  if (holder == null) {
+                    throw new RuntimeException(new TimeoutException());
+                  }
+                  nextVal = holder.val;
+                  return nextVal != null;
+                }
+                catch (InterruptedException e) {
+                  throw new RuntimeException(e);
```

Comment:
Would you tell me why? Is there anything else which should handle `InterruptedException` particularly?

Reply:
Everywhere that catches Exception that could be an interrupted exception, or anywhere that catches InterruptedException will swallow the interrupted flag on the thread. You can search the code for usages of `java.lang.Thread#interrupt` for other places it is done. It is a nasty latent bug source since it can easily result in an interrupt being lost and a retry loop clogging up a shutdown or recovery cycle of some kind. There's a reason the scala includes InterruptedException in its [fatal](https://www.scala-lang.org/api/2.9.3/scala/util/control/NonFatal$.html) list of exceptions.

Reply:
I know. My question is what's the point of interrupting itself again? This exception is just thrown by the threads who run the runnable in background.